File: IPB.R
set.seed(42)

# -----------------------------
# 1. Generate GPD data manually
# -----------------------------
rgpd_manual <- function(n, xi, sigma) {
  u <- runif(n)
  
  if (abs(xi) > 1e-6) {
    return(sigma * ((u^(-xi) - 1) / xi))
  } else {
    return(-sigma * log(u))  # exponential case
  }
}

true_xi <- 0.2
true_sigma <- 1.5

data <- rgpd_manual(500, true_xi, true_sigma)


# -----------------------------
# 2. Log-likelihood (manual)
# -----------------------------
log_likelihood <- function(xi, sigma, data) {
  
  if (sigma <= 0) return(-Inf)
  
  # constraint: 1 + xi*x/sigma > 0
  if (any(1 + xi * data / sigma <= 0)) return(-Inf)
  
  n <- length(data)
  
  if (abs(xi) > 1e-6) {
    ll <- -n * log(sigma) - (1/xi + 1) *
      sum(log(1 + xi * data / sigma))
  } else {
    ll <- -n * log(sigma) - sum(data) / sigma
  }
  
  return(ll)
}


# -----------------------------
# 3. Informative priors
# -----------------------------
log_prior <- function(xi, sigma) {
  
  if (sigma <= 0) return(-Inf)
  
  xi_prior <- -0.5 * ((xi - 0.3) / 0.3)^2
  sigma_prior <- -0.5 * ((sigma - 2) / 1)^2
  
  return(xi_prior + sigma_prior)
}


# -----------------------------
# 4. Posterior
# -----------------------------
log_posterior <- function(xi, sigma, data) {
  log_likelihood(xi, sigma, data) + log_prior(xi, sigma)
}


# -----------------------------
# 5. Metropolis-Hastings
# -----------------------------
iterations <- 10000

xi_chain <- numeric(iterations)
sigma_chain <- numeric(iterations)

xi_chain[1] <- 0.5
sigma_chain[1] <- 2.5

for (j in 2:iterations) {
  
  xi_star <- rnorm(1, xi_chain[j-1], 0.05)
  sigma_star <- rnorm(1, sigma_chain[j-1], 0.1)
  
  log_r <- log_posterior(xi_star, sigma_star, data) -
           log_posterior(xi_chain[j-1], sigma_chain[j-1], data)
  
  if (log(runif(1)) < log_r) {
    xi_chain[j] <- xi_star
    sigma_chain[j] <- sigma_star
  } else {
    xi_chain[j] <- xi_chain[j-1]
    sigma_chain[j] <- sigma_chain[j-1]
  }
}


# -----------------------------
# 6. Estimates
# -----------------------------
burn_in <- 2000

xi_estimate <- mean(xi_chain[burn_in:iterations])
sigma_estimate <- mean(sigma_chain[burn_in:iterations])

cat("Estimated xi:", xi_estimate, "\n")
cat("Estimated sigma:", sigma_estimate, "\n")

File: ipb.py
import numpy as np
from scipy.stats import genpareto
import matplotlib.pyplot as plt

# -----------------------------
# 1. Generate fake extreme loss data
# -----------------------------
np.random.seed(42)

true_xi = 0.2      # true tail parameter
true_sigma = 1.5  # true scale

data = genpareto.rvs(c=true_xi, scale=true_sigma, size=500)


# -----------------------------
# 2. Likelihood function
# -----------------------------
def log_likelihood(xi, sigma, data):
    # parameters must be positive
    if sigma <= 0:
        return -np.inf
    
    return np.sum(genpareto.logpdf(data, c=xi, scale=sigma))


# -----------------------------
# 3. Prior (informative prior)
# assume xi ~ N(0.3, 0.3)
# assume sigma ~ N(2, 1)
# -----------------------------
def log_prior(xi, sigma):
    if sigma <= 0:
        return -np.inf
    
    xi_prior = -0.5 * ((xi - 0.3) / 0.3)**2
    sigma_prior = -0.5 * ((sigma - 2) / 1)**2
    
    return xi_prior + sigma_prior


# -----------------------------
# 4. Posterior = likelihood + prior
# -----------------------------
def log_posterior(xi, sigma, data):
    return log_likelihood(xi, sigma, data) + log_prior(xi, sigma)


# -----------------------------
# 5. Metropolis-Hastings
# -----------------------------
iterations = 10000

xi_chain = []
sigma_chain = []

# starting guess
current_xi = 0.5
current_sigma = 2.5

for i in range(iterations):
    
    # propose new candidate
    proposal_xi = np.random.normal(current_xi, 0.05)
    proposal_sigma = np.random.normal(current_sigma, 0.1)
    
    # compute acceptance ratio
    current_post = log_posterior(current_xi, current_sigma, data)
    proposal_post = log_posterior(proposal_xi, proposal_sigma, data)
    
    acceptance_ratio = np.exp(proposal_post - current_post)
    
    # accept/reject
    if np.random.rand() < acceptance_ratio:
        current_xi = proposal_xi
        current_sigma = proposal_sigma
    
    xi_chain.append(current_xi)
    sigma_chain.append(current_sigma)


# -----------------------------
# 6. Results
# -----------------------------
print("Estimated xi:", np.mean(xi_chain[2000:]))      # burn-in removed
print("Estimated sigma:", np.mean(sigma_chain[2000:]))

# Plot chains
plt.figure(figsize=(12,5))

plt.subplot(1,2,1)
plt.plot(xi_chain)
plt.title("Xi Chain")

plt.subplot(1,2,2)
plt.plot(sigma_chain)
plt.title("Sigma Chain")

plt.show()
